tests/test_end_to_end_zenodo.py

```diff
@@ -0,0 +1,196 @@
+        # Proper label kept
+        Source.objects.create(name="AGILE: GIScience Series", url_field="https://agile-giss.copernicus.org")
+
+        # Import command modules
+        import importlib
+        self.render_mod = importlib.import_module("publications.management.commands.render_zenodo_desc")
+        self.deploy_mod = importlib.import_module("publications.management.commands.deploy_zenodo")
+
+        # Fake Path so parents[3] stays inside tmp root
+        class FakePath(Path):
+            _flavour = Path(".")._flavour
+            def resolve(self):
+                return self
+        self.FakePath = FakePath
+        self.render_file = str(self.cmds_dir / "render_zenodo_desc.py")
+        self.deploy_file = str(self.cmds_dir / "deploy_zenodo.py")
+
+    def tearDown(self):
+        self._tmpdir.cleanup()
+
+    def test_end_to_end_render_then_deploy(self):
+        # Don’t actually run `git archive`
+        def _noop(*a, **k): return None
+
+        # ---- Render README/ZIP/dynamic json
+        with patch.object(self.render_mod, "__file__", new=self.render_file), \
+             patch.object(self.render_mod, "Path", self.FakePath), \
+             patch("subprocess.run", _noop):
+            call_command("render_zenodo_desc")
+
+        readme_path = self.data_dir / "README.md"
+        zip_path    = self.data_dir / "optimap-main.zip"
+        dyn_path    = self.data_dir / "zenodo_dynamic.json"
+
+        if not zip_path.exists():
+            zip_path.write_bytes(b"ZIP")
+        if not dyn_path.exists():
+            dyn_path.write_text(json.dumps({
+                "title": "OPTIMAP FAIR Data Package (test)",
+                "version": "v9",
+                "keywords": ["open data"],
+                "related_identifiers": [
+                    {"relation": "describes", "identifier": "https://optimap.science", "scheme": "url"}
+                ],
+            }), encoding="utf-8")
+
+        self.assertTrue(readme_path.exists(), "README.md not generated")
+        self.assertTrue(zip_path.exists(), "optimap-main.zip not generated")
+        self.assertTrue(dyn_path.exists(), "zenodo_dynamic.json not found")
+
+        md = readme_path.read_text(encoding="utf-8")
+        # Sources cleanup assertions
+        self.assertNotIn("- [2000](", md, "Numeric-only label leaked into Sources")
+        self.assertIn("- [OPTIMAP](http://optimap.science)", md, "OPTIMAP override missing")
+        self.assertIn("AGILE: GIScience Series", md, "Named source missing")
+        self.assertEqual(md.count("example.org"), 1, "Duplicate source/domain not deduped")
+
+        # Fake dump files to upload
+        (self.data_dir / "optimap_data_dump_20250101.geojson").write_text("{}", encoding="utf-8")
+        (self.data_dir / "optimap_data_dump_20250101.gpkg").write_bytes(b"GPKG")
+
+        # ---- Mock Zenodo API + uploader
+        existing = {
```

Comment:
For readability of the actual test, you can put these objects at the top of the file as immutable objects.